Add permissions boundary policy to Terraformer role

dav3r opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR creates a [permissions boundary policy](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_boundaries.html) and applies it to the Terraformer role.

The rules of the Terraformer boundary policy are:
* Not allowed to modify any resources tagged by the team that deploys the assessment environment (aside from a small number of exceptions that have been granted)
* Not allowed to assume the Guacamole, Samba, and Terraformer instance roles
* Not allowed to modify CloudFormation resources created by Control Tower
* Not allowed to modify or delete the permissions boundary policy
* Not allowed to remove the permissions boundary policy from users and roles that have it
* Not allowed to create users or roles that do not include the permissions boundary
* Not allowed to apply any other permission boundary policies to users or roles

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

Previously, no IAM permissions were allowed by the Terraformer role because that would've allowed the creation of a policy with admin access.  This change allows the Terraformer role to perform IAM functions (e.g. creating custom roles and policies) as long as they don't break the rules specified in the permissions boundary policy.  This will make the Terraformer role more useful to our users and enable them to deploy a wider variety of resources in their environments.

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

I deployed these changes to a test environment and attempted to perform activities that should be prohibited by the boundary policy (e.g. modify resources that were tagged by deployment team, replace the boundary policy with a different one, remove the boundary policy from the Terraformer role).  In every case that I tested, I was unable to violate the rules of the boundary policy.

In addition, I confirmed (with help from others) that this boundary policy allows the following:
- [x] Successful deployment of [NuclearPond](https://github.com/DevSecOpsDocs/terraform-nuclear-pond) (with minor changes to the standard NuclearPond Terraform)
- [x] Successful deployment of all Terraform used by the Advanced Operations team

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

dav3r committed:
Add variables for Terraformer permissions boundary policy

dav3r committed:
Define Terraformer boundary policy

This policy allows full AWS permissions except for the following:
* Deny modification of any resources tagged by the team that deploys this root module, with some exceptions (detailed below).
* Deny use of the Guacamole, Samba, and Terraformer instance roles.
* Deny modification or deletion of this permissions boundary policy.
* Deny removal of this permissions boundary policy from users and roles.
* Deny creation of users or roles that do not have this permissions boundary.
* Deny applying any other permission boundary policies to users or roles.

dav3r committed:
Update the Terraformer policy and role to use permissions boundary policy:

* Attach permission boundary policy to Terraformer role
* Remove IAM and other restrictions from Terraformer policy (replaced by boundary policy restrictions)
* Remove unneeded "Allow" statements from Terraformer policy (since effect = "Allow" is the default)
* Remove now-unneeded "Read Only" policy attachment from Terraformer role
* Update comments accordingly

dav3r committed:
Update Terraform documentation in README

dav3r committed:
Add a missing bit of info to boundary policy comment

Also correct spelling of "Control Tower" in a different comment.

jsf9k commented on variables.tf:
```diff
@@ -330,6 +330,18 @@ variable "tags" {
+variable "terraformer_permissions_boundary_policy_description" {
+  type        = string
+  description = "The description to associate with the IAM permissions boundary policy attached to the Terraformer instance role in order to protect the foundational resources deployed in this account."
+  default     = "Defines the IAM permissions boundary for Terraformer instances in order to protect the foundational resources deployed in this account."
```
I think this language from the `description` field is clearer and can simply be repeated:
```suggestion
  default     = "The IAM permissions boundary policy attached to the Terraformer instance role in order to protect the foundational resources deployed in this account."
```

jsf9k commented on terraformer_policy.tf:
```diff
@@ -1,70 +1,32 @@
-    ]
-  }
 
   # Allow use of the KMS key used to encrypt COOL AMIs.
```
I think we should have the same comments for each of these `statement`s as are used in the corresponding `statement`s in `terraformer_permissionsboundary.tf`.  The latter are more informative, and it's the same code.

dav3r replied:
Good call - updated in https://github.com/cisagov/cool-assessment-terraform/pull/215/commits/df75069c286cd745eac69f96cc66d5cd38772bb0.

jsf9k reviewed (approved):
Very strong work here, @dav3r.  @mcdonnnj was definitely on the right path with his suggestion.

I have a couple of suggestions for your consideration, but if you disagree I still approve the PR.

dav3r committed:
Make comment verbiage consistent between Terraformer policy and boundary policy

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

dav3r committed:
Improve default wording of Terraformer permissions boundary policy description

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

dav3r committed:
Update Terraformer permissions boundary

Allow read-only access to EC2 resources tagged by the team that deploys this root module.

Also, remove a permission made redundant by the change above.

dav3r committed:
Update Terraformer role and policy

Allow read-only access to everything in AWS (to be limited by the permissions boundary policy).  This is specifically to allow read access to existing resources that _are_ tagged as being created by the team that deploys this root module.

Also, remove a permission made redundant by the change above.

dav3r committed:
Update Terraform resource documentation in README

dav3r committed:
Add Terraformer permissions boundary policy to outputs

jsf9k reviewed (approved):
:heart_eyes_cat:

mcdonnnj commented on terraformer_permissionsboundary.tf:
```diff
@@ -0,0 +1,291 @@
+    condition {
+      test = "StringEquals"
+      values = [
+        var.tags["Team"],
```
This variable is optional. If it is not provided, or the key is absent from the provided value, I believe this will break. I think `lookup()` (with a default value) is the only way to safely access a possibly undefined value.

dav3r replied:
Excellent catch!  Please let me know if you're good with https://github.com/cisagov/cool-assessment-terraform/pull/215/commits/9b061a379b7d598125e40e8a711414ed270ddff5.

mcdonnnj commented on terraformer_permissionsboundary.tf:
```diff
@@ -0,0 +1,291 @@
+    condition {
+      test = "StringNotEquals"
+      values = [
+        var.tags["Team"],
```
Same comment about unsafe key access.

dav3r replied:
See https://github.com/cisagov/cool-assessment-terraform/pull/215/commits/9b061a379b7d598125e40e8a711414ed270ddff5.

mcdonnnj commented on terraformer_policy.tf:
```diff
@@ -1,72 +1,35 @@
     condition {
       test = "StringNotEquals"
       values = [
         var.tags["Team"],
```
Guess I missed this in previous reviews but same question about unsafe access.

dav3r replied:
See https://github.com/cisagov/cool-assessment-terraform/pull/215/commits/9b061a379b7d598125e40e8a711414ed270ddff5.

mcdonnnj commented on terraformer_policy.tf:
```diff
@@ -1,72 +1,35 @@
-  # IAM access to be as read-only as possible in order to stop the
-  # Terraformer instance from defeating the Terraformer policy by
-  # creating new users, policies, roles, etc.
+  # Deny modification of any resources tagged by the team that deploys this
```
I'm not a fan of this phrasing because there's no explicit `Deny` going on here. I would rather frame it terms of what is allowed since this is an allow statement.

dav3r replied:
Please take a look at https://github.com/cisagov/cool-assessment-terraform/pull/215/commits/b7a9b68396c9557f73bcc6601cff39e4bc64f282 and let me know what you think.

dav3r committed:
Handle case where Team tag is not present

If the Team tag is not present, use a default value that provides some information to the user.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Rephrase comment in terms of allow vs. deny

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj commented on terraformer_permissionsboundary.tf:
```diff
@@ -0,0 +1,291 @@
+    sid = "AllowUsingCOOLAMIKMSKey"
+  }
+
+  # Allow the launching of new instances in the operations subnet,
```
This also allows launching in the private subnet no?

dav3r replied:
You are correct - good catch, see https://github.com/cisagov/cool-assessment-terraform/pull/215/commits/39690032e1aff1d560d939d6cf3b7a171479e366.

mcdonnnj commented on terraformer_permissionsboundary.tf:
```diff
@@ -0,0 +1,291 @@
+    sid = "AllowModifyingS3Endpoint"
+  }
+
+  # Don't allow Terraformer instances to touch the CloudFormation foo
+  # put in place by Control Tower.  This is not covered by the earlier
+  # statement allowing full access to resources that are not tagged as
+  # belonging to the team that deploys this root module, since CloudFormation
+  # resources do not accept tags.
```
Maybe I'm just having a time while reading this comment, but it starts by talking about what's being denied and then says we're doing it because it's not covered by an allow?

dav3r replied:
Take a look below at my attempt to make this read better:

```
  # Don't allow Terraformer instances to touch the CloudFormation foo put in
  # place by Control Tower.  CloudFormation resources do not accept tags and
  # thus cannot be tagged with a team tag, which is why this statement is needed
  # (an earlier statement prevents modification of any resources tagged by the
  # team that deploys this root module).
```

I'm open to suggestions if this doesn't work for you.

mcdonnnj replied:
That works for me. 👍

dav3r replied:
Cool, please see https://github.com/cisagov/cool-assessment-terraform/pull/215/commits/7adcc68974453e1d9de2f9fe7a180c723d27341e.

mcdonnnj reviewed (approved):
Two small questions about a couple comments but otherwise LGTM ✔

dav3r committed:
Make a comment more accurate

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Improve readability of a comment

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>
